oxide_core/quaternion.rs

```diff
@@ -0,0 +1,169 @@
+use crate::vector::vector3::*;
+use crate::vector::*;
+use std::ops;
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub struct Quaternion {
+  pub x: f32,
+  pub y: f32,
+  pub z: f32,
+  pub w: f32,
+}
+
+pub trait VectorPart {
+  fn get_vector_part(&self) -> Vector3;
+}
+
+impl VectorPart for Quaternion {
+  fn get_vector_part(&self) -> Vector3 {
+    Vector3 {
+      x: self.x,
+      y: self.y,
+      z: self.z,
+    }
+  }
+}
+
+pub trait ScalarPart {
+  fn get_scalar_part(&self) -> f32;
+}
+
+impl ScalarPart for Quaternion {
+  fn get_scalar_part(&self) -> f32 {
+    self.w
+  }
+}
+
+pub fn id() -> Quaternion {
+  let one = 1.0f32;
+  let zero = 0.0f32;
+
+  Quaternion {
+    x: zero,
+    y: zero,
+    z: zero,
+    w: one,
+  }
+}
+
+impl Scale for Quaternion {
+  fn scale(&self, t: f32) -> Self {
+    Quaternion {
+      x: self.x * t,
+      y: self.y * t,
+      z: self.z * t,
+      w: self.w * t,
+    }
+  }
+}
+
+pub fn dot(a: Quaternion, b: Quaternion) -> f32 {
+  a.get_scalar_part() * b.get_scalar_part()
+    + a.get_vector_part().inner_product(&b.get_vector_part())
+}
+
+pub fn inverse(q: Quaternion) -> Quaternion {
+  q.scale(q.get_length().recip())
+}
+
+pub fn conjugate(a: Quaternion) -> Quaternion {
+  let vector_neg = a.get_vector_part().negate();
+  Quaternion {
+    x: vector_neg.x,
+    y: vector_neg.y,
+    z: vector_neg.z,
+    w: a.w,
+  }
+}
+
+impl Quaternion {
+  pub fn get_squared_length(&self) -> f32 {
+    self.w * self.w + self.get_vector_part().get_squared_length()
+  }
+  pub fn get_length(&self) -> f32 {
+    self.get_squared_length().sqrt()
+  }
+}
```

Comment:
I think that these functions don't need.

Reply:
I think so. I'll remove it.

Reply:
✅ [`02d4435` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/02d4435b02e47cc1c162a13c47ef5aead1cdcbc4)

Reply:
I noticed that function ```get_length``` is needed to implement ```inverse``` function.
So, I changed the visibility of those two functions.
✅ [`bce2141` (#14)](https://github.com/utilForever/Oxide/pull/14/commits/bce214119efc72065ee66d8ddc01cdc8c11c529f)